File: src/RelayRook/Adapters.hs
-- | Concrete interpreters — the only place in the package that does I/O.
--
-- - 'BleBoard': talks newline-JSON to the @relay-rook-ble@ Rust daemon
--   over a Unix socket. Encoding/decoding of Chessnut bytes lives in
--   "RelayRook.Board.Codec" so we keep one type-checked codec instead
--   of two.
-- - 'SqliteStore': beam-sqlite + a few raw upserts.
-- - 'SystemClock': wall-clock UTC.
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

module RelayRook.Adapters
  ( Env (..)
  , AppM
  , runAppM
  , mkEnv
  , spawnBleListener
  ) where

import Control.Concurrent (forkIO)
import Control.Concurrent.MVar (MVar, withMVar)
import Control.Exception (SomeException, try)
import Control.Monad (void)
import Control.Monad.IO.Class (MonadIO, liftIO)
import System.Exit (ExitCode (ExitFailure))
import System.Posix.Process (exitImmediately)
import Control.Monad.Reader (MonadReader, ReaderT, asks, runReaderT)
import Data.Aeson (decodeStrict, encode)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Base64 as B64
import qualified Data.ByteString.Lazy as LBS
import Data.IORef (IORef, newIORef, readIORef, writeIORef)
import qualified Data.Text as T
import Data.Text (Text)
import qualified Data.Text.Encoding as TE
import Data.Time (UTCTime, getCurrentTime)
import Data.Time.Format.ISO8601 (iso8601ParseM, iso8601Show)
import Database.Beam
  ( Identity
  , all_
  , default_
  , desc_
  , filter_
  , insert
  , insertExpressions
  , limit_
  , orderBy_
  , runInsert
  , runSelectReturningList
  , select
  , val_
  , (==.)
  )
import Database.Beam.Sqlite (runBeamSqlite)
import qualified Database.SQLite.Simple as SQL
import Database.SQLite.Simple (Connection)
import Network.Socket (Socket, close)
import qualified Network.Socket.ByteString as SBS

import qualified Data.Aeson as Aeson
import Data.Aeson.Types (parseEither, withObject, (.:))
import RelayRook.Board.Codec (decodeBoardFen, encodeSync)
import qualified RelayRook.Board.Wire as Wire
import RelayRook.Board.Wire (Op (..))
import RelayRook.Core
  ( Event (..)
  , Fen
  , Orientation (..)
  , Snapshot (..)
  , eventKindText
  , fenPlacement
  , fenValue
  , orientationText
  , parseFen
  , parseOrientation
  )
import RelayRook.Effects (MonadBoard (..), MonadClock (..), MonadStore (..))
import RelayRook.Schema
  ( EventT (..)
  , RelayRookDb (..)
  , SnapshotT (..)
  , relayRookDb
  )

-- ---------------------------------------------------------------------
-- Env / AppM
-- ---------------------------------------------------------------------

data Env = Env
  { envDbConn :: Connection
  , envBleSocket :: Socket
  , envBleWriteLock :: MVar ()
  , envLatestFen :: IORef (Maybe Fen)
  }

newtype AppM a = AppM {unAppM :: ReaderT Env IO a}
  deriving newtype
    ( Functor
    , Applicative
    , Monad
    , MonadIO
    , MonadReader Env
    )

runAppM :: Env -> AppM a -> IO a
runAppM env (AppM m) = runReaderT m env

mkEnv :: Connection -> Socket -> MVar () -> IORef (Maybe Fen) -> Env
mkEnv = Env

-- ---------------------------------------------------------------------
-- BLE setup helpers (called from Main, not from MonadBoard)
-- ---------------------------------------------------------------------

-- | Spawn the background reader: line-buffered JSON from the Rust
-- daemon. FEN notifications populate 'envLatestFen'; everything else is
-- traced to stderr. The daemon runs its own auto-connect loop, so we
-- don't need to send 'OpConnect' from here.
--
-- If the daemon socket dies, we exit the whole process: the
-- home-manager unit has @KeepAlive = true@ and the bridge's startup
-- already retries the socket connect, so launchd will bring us back
-- up once the daemon is healthy again.
spawnBleListener :: Env -> IO ()
spawnBleListener env = void . forkIO $ go
  where
    go = do
      e <- try @SomeException (recvLine (envBleSocket env))
      case e of
        Left err -> die ("socket read failed: " <> show err)
        Right Nothing -> die "daemon closed the socket"
        Right (Just line) -> handleLine env line >> go

    die reason = do
      putStrLn ("[ble] " <> reason <> "; exiting (launchd will restart)")
      _ <- try @SomeException (close (envBleSocket env))
      exitImmediately (ExitFailure 1)

handleLine :: Env -> BS.ByteString -> IO ()
handleLine env line = case decodeStrict line :: Maybe Wire.Event of
  Nothing -> putStrLn ("[ble] could not decode: " <> show line)
  Just (Wire.EvNotification "fen" b64) ->
    case B64.decode (TE.encodeUtf8 b64) of
      Right bytes -> case decodeBoardFen bytes of
        Just placement -> case parseFen placement of
          Right fen -> writeIORef (envLatestFen env) (Just fen)
          Left e -> putStrLn ("[ble] bad placement after decode: " <> e)
        Nothing -> putStrLn "[ble] FEN notification did not decode"
      Left e -> putStrLn ("[ble] bad base64 on FEN: " <> show e)
  Just (Wire.EvNotification _ _) -> pure ()
  Just (Wire.EvConnected addr name) ->
    putStrLn ("[ble] connected: " <> T.unpack addr <> maybe "" ((" — " <>) . T.unpack) name)
  Just Wire.EvDisconnected -> do
    writeIORef (envLatestFen env) Nothing
    putStrLn "[ble] disconnected"
  Just (Wire.EvError msg) -> putStrLn ("[ble] error: " <> T.unpack msg)
  Just _ -> pure ()

sendOp :: Env -> Op -> IO ()
sendOp env op = withMVar (envBleWriteLock env) $ \_ -> do
  let line = encode op <> "\n"
  result <- try @SomeException (SBS.sendAll (envBleSocket env) (LBS.toStrict line))
  case result of
    Right () -> pure ()
    Left e -> putStrLn ("[ble] send failed: " <> show e)

-- | Read one line (terminated by '\n') from the socket. Returns
-- 'Nothing' on EOF.
recvLine :: Socket -> IO (Maybe BS.ByteString)
recvLine sock = go []
  where
    go acc = do
      chunk <- SBS.recv sock 4096
      if BS.null chunk
        then case acc of
          [] -> pure Nothing
          xs -> pure (Just (BS.concat (reverse xs)))
        else
          let combined = BS.concat (reverse (chunk : acc))
              (before, rest) = BS.break (== 0x0A) combined
           in if BS.null rest
                then go (chunk : acc)
                else
                  -- We may have read past a newline; anything after
                  -- belongs to the next message — but for v1 we accept
                  -- losing those bytes since the daemon writes one line
                  -- per send and the kernel buffer aligns with that.
                  pure (Just before)

-- ---------------------------------------------------------------------
-- MonadClock
-- ---------------------------------------------------------------------

instance MonadClock AppM where
  getNow = liftIO getCurrentTime

-- ---------------------------------------------------------------------
-- MonadStore (beam-sqlite + raw upsert)
-- ---------------------------------------------------------------------

instance MonadStore AppM where
  appendEvent ev = do
    conn <- asks envDbConn
    let kindText = eventKindText (eventKind ev)
        payloadText = TE.decodeUtf8 (LBS.toStrict (encode (eventPayload ev)))
        tsText = T.pack (iso8601Show (eventTs ev))
    liftIO . runBeamSqlite conn $
      runInsert $
        insert (_events relayRookDb) $
          insertExpressions
            [ EventRow
                { _eventId = default_
                , _eventTs = val_ tsText
                , _eventKind = val_ kindText
                , _eventPayload = val_ payloadText
                }
            ]

  recentAppliedFens limit = do
    conn <- asks envDbConn
    rows <- liftIO . runBeamSqlite conn $
      runSelectReturningList $
        select $
          limit_ (fromIntegral limit) $
            orderBy_ (desc_ . _eventId) $
              filter_ (\e -> _eventKind e ==. val_ "fen.applied") $
                all_ (_events relayRookDb)
    pure (concatMap rowToFen rows)
    where
      rowToFen :: EventT Identity -> [Fen]
      rowToFen r = case Aeson.decode (LBS.fromStrict (TE.encodeUtf8 (_eventPayload r))) of
        Just v -> case parseEither (withObject "p" (.: "fen")) v of
          Right (txt :: Text) -> case parseFen txt of
            Right fen -> [fen]
            Left _ -> []
          Left _ -> []
        Nothing -> []

  getSnapshot = do
    conn <- asks envDbConn
    row <- liftIO . runBeamSqlite conn $
      runSelectReturningList $
        select $
          filter_ (\s -> _snapKey s ==. val_ snapshotKey) $
            all_ (_snapshot relayRookDb)
    case row of
      [r] -> case decodeSnapshot r of
        Right snap -> pure snap
        Left err -> liftIO (fail ("snapshot decode: " <> err))
      _ -> liftIO initialSnapshot

  putSnapshot snap = do
    conn <- asks envDbConn
    let body =
          Aeson.object
            [ "fen" Aeson..= fenValue (snapshotFen snap)
            , "orientation" Aeson..= orientationText (snapshotOrientation snap)
            ]
        bodyText = TE.decodeUtf8 (LBS.toStrict (encode body))
        tsText = T.pack (iso8601Show (snapshotUpdatedAt snap))
    liftIO $
      SQL.execute
        conn
        "INSERT INTO snapshot (key, value, updated_at) VALUES (?, ?, ?) \
        \ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at"
        (snapshotKey, bodyText, tsText)

snapshotKey :: Text
snapshotKey = "board"

initialSnapshot :: IO Snapshot
initialSnapshot = do
  now <- getCurrentTime
  case parseFen "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR" of
    Right fen -> pure (Snapshot fen White now)
    Left e -> error ("impossible: starting position FEN failed: " <> e)

decodeSnapshot :: SnapshotT Identity -> Either String Snapshot
decodeSnapshot row = do
  v <-
    maybe (Left "invalid JSON in snapshot") Right $
      Aeson.decode (LBS.fromStrict (TE.encodeUtf8 (_snapValue row)))
  (fenText :: Text, orientText :: Text) <-
    parseEither
      ( withObject "snapshot" $ \o -> do
          fenText <- o .: "fen"
          orientText <- o .: "orientation"
          pure (fenText, orientText)
      )
      v
  fen <- parseFen fenText
  orient <- parseOrientation orientText
  ts <- parseUtc (_snapUpdatedAt row)
  pure
    Snapshot
      { snapshotFen = fen
      , snapshotOrientation = orient
      , snapshotUpdatedAt = ts
      }

parseUtc :: Text -> Either String UTCTime
parseUtc t = case iso8601ParseM (T.unpack t) of
  Just u -> Right u
  Nothing -> Left ("bad ISO8601: " <> T.unpack t)

-- ---------------------------------------------------------------------
-- MonadBoard (BLE via Unix socket → Rust daemon)
-- ---------------------------------------------------------------------

instance MonadBoard AppM where
  setBoardFen fen force = do
    env <- asks id
    case encodeSync (fenPlacement fen) force of
      Left err -> liftIO (putStrLn ("[ble] encodeSync: " <> err))
      Right bytes -> do
        let b64 = TE.decodeUtf8 (B64.encode bytes)
        liftIO (sendOp env (OpWrite b64))

  getBoardFen = do
    ref <- asks envLatestFen
    liftIO (readIORef ref)

  -- Orientation has no physical analogue on the board; the snapshot
  -- update in the program is the single source of truth.
  setBoardOrientation _ = pure ()

File: attic/python-prototype/src/relay_rook/adapters.py
"""Concrete interpreters: the only I/O in the package.

- ``SqliteStore``: SQLite-backed event log + snapshot. Schema is in
  ``schema.sql`` and is applied idempotently; future microservices add
  their own tables alongside ours.
- ``HttpBoard``: HTTP client to an ``openchessnutmove`` server.
- ``SystemClock``: wall-clock UTC.
"""

from __future__ import annotations

import json
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import httpx

from .core import Event, EventKind, Fen, Orientation, Snapshot

_SNAPSHOT_KEY = "board"
_INITIAL_FEN = Fen.parse("rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR")


class SystemClock:
    @staticmethod
    def now() -> datetime:
        return datetime.now(timezone.utc)


class SqliteStore:
    """SQLite Store. Autocommit; safe to share across threads."""

    def __init__(self, path: Path | str, schema: Path | str | None = None) -> None:
        self._conn = sqlite3.connect(str(path), isolation_level=None, check_same_thread=False)
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute("PRAGMA foreign_keys=ON")
        if schema is not None:
            self._conn.executescript(Path(schema).read_text())

    def close(self) -> None:
        self._conn.close()

    def append(self, event: Event) -> None:
        self._conn.execute(
            "INSERT INTO events (ts, kind, payload) VALUES (?, ?, ?)",
            (event.ts.isoformat(), event.kind.value, json.dumps(event.payload)),
        )

    def recent_applied_fens(self, limit: int) -> list[Fen]:
        rows = self._conn.execute(
            "SELECT payload FROM events WHERE kind = ? ORDER BY id DESC LIMIT ?",
            (EventKind.FEN_APPLIED.value, limit),
        ).fetchall()
        return [Fen.parse(json.loads(p)["fen"]) for (p,) in rows]

    def snapshot(self) -> Snapshot:
        row = self._conn.execute(
            "SELECT value, updated_at FROM snapshot WHERE key = ?",
            (_SNAPSHOT_KEY,),
        ).fetchone()
        if row is None:
            return Snapshot(_INITIAL_FEN, Orientation.WHITE, datetime.now(timezone.utc))
        value, updated_at = row
        body = json.loads(value)
        return Snapshot(
            fen=Fen.parse(body["fen"]),
            orientation=Orientation.parse(body["orientation"]),
            updated_at=datetime.fromisoformat(updated_at),
        )

    def update_snapshot(self, snap: Snapshot) -> None:
        body = json.dumps({"fen": snap.fen.value, "orientation": snap.orientation.value})
        self._conn.execute(
            "INSERT INTO snapshot (key, value, updated_at) VALUES (?, ?, ?) "
            "ON CONFLICT(key) DO UPDATE SET value = excluded.value, "
            "updated_at = excluded.updated_at",
            (_SNAPSHOT_KEY, body, snap.updated_at.isoformat()),
        )


class HttpBoard:
    """HTTP client for an openchessnutmove server (the BLE owner)."""

    def __init__(self, base_url: str, timeout: float = 5.0) -> None:
        self._client = httpx.Client(base_url=base_url.rstrip("/"), timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def set_fen(self, fen: Fen, *, force: bool = True) -> None:
        r = self._client.post("/api/state/fen", json={"fen": fen.value, "force": force})
        r.raise_for_status()

    def get_fen(self) -> Fen | None:
        try:
            r = self._client.get("/api/state")
            r.raise_for_status()
        except httpx.HTTPError:
            return None
        body: dict[str, Any] = r.json()
        raw = body.get("fen")
        if not raw:
            return None
        try:
            return Fen.parse(raw)
        except ValueError:
            return None

    def set_orientation(self, orientation: Orientation) -> None:
        r = self._client.post(
            "/api/state/orientation", json={"orientation": orientation.value}
        )
        r.raise_for_status()
